### ISSUE TITLE
Issues cross-compiling clFFT

### ISSUE THREAD

**Issue by juliohm** [clMathLibraries/clFFT#243]
I am trying to cross-compile clFFT for all supported platforms, to make it available for Julia users. I am seeing various warnings that I would like to share with the maintainers:

```
┌ Warning: /home/juliohm/Desktop/cross-platform-clfft/build/x86_64-linux-musl/zWHvgpfW/x86_64-linux-musl-libgfortran3-cxx11/destdir/lib64/libclFFT.so.2.12.2 contains std::string values!  This causes incompatibilities across the GCC 4/5 version boundary.

┌ Warning: Minimum instruction set detected for lib64/libclFFT.so.2.12.2 is avx2, not x86_64 as desired.
```

The cross-compilation script is here in case you want to reproduce: https://github.com/juliohm/cross-platform-clfft

Can you please confirm the list of supported platforms?

**Comment by juliohm** [comment]
Does clFFT work on 32-bit platforms?

```
$ julia --project build_tarballs.jl i686-linux-gnu                                                                    ✔ 
[ Info: Building for i686-linux-gnu
-- Found FFTW: /opt/i686-linux-gnu/i686-linux-gnu/sys-root/usr/local/lib/libfftw3f.so;/opt/i686-linux-gnu/i686-linux-gnu/sys-root/usr/local/lib/libfftw3.so  
-- Detected GNU fortran compiler.
-- CMAKE_CXX_COMPILER flags: -m64 -pthread 
-- CMAKE_CXX_COMPILER debug flags: -g
-- CMAKE_CXX_COMPILER release flags: -O3 -DNDEBUG
-- CMAKE_CXX_COMPILER relwithdebinfo flags: -O2 -g -DNDEBUG
-- CMAKE_EXE_LINKER link flags: 
FFT clients will NOT be built
GoogleTest unit tests will NOT be built
FFT callback client will NOT be built
-- Configuring done
-- Generating done
-- Build files have been written to: /workspace/srcdir/clFFT/build
/usr/bin/cmake -S/workspace/srcdir/clFFT/src -B/workspace/srcdir/clFFT/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /workspace/srcdir/clFFT/build/CMakeFiles /workspace/srcdir/clFFT/build//CMakeFiles/progress.marks
/usr/bin/make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/workspace/srcdir/clFFT/build'
/usr/bin/make  -f library/CMakeFiles/clFFT.dir/build.make library/CMakeFiles/clFFT.dir/depend
/usr/bin/make  -f statTimer/CMakeFiles/StatTimer.dir/build.make statTimer/CMakeFiles/StatTimer.dir/depend
make[2]: Entering directory '/workspace/srcdir/clFFT/build'
cd /workspace/srcdir/clFFT/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /workspace/srcdir/clFFT/src /workspace/srcdir/clFFT/src/library /workspace/srcdir/clFFT/build /workspace/srcdir/clFFT/build/library /workspace/srcdir/clFFT/build/library/CMakeFiles/clFFT.dir/DependInfo.cmake --color=
make[2]: Entering directory '/workspace/srcdir/clFFT/build'
cd /workspace/srcdir/clFFT/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /workspace/srcdir/clFFT/src /workspace/srcdir/clFFT/src/statTimer /workspace/srcdir/clFFT/build /workspace/srcdir/clFFT/build/statTimer /workspace/srcdir/clFFT/build/statTimer/CMakeFiles/StatTimer.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/workspace/srcdir/clFFT/build'
make[2]: Leaving directory '/workspace/srcdir/clFFT/build'
/usr/bin/make  -f statTimer/CMakeFiles/StatTimer.dir/build.make statTimer/CMakeFiles/StatTimer.dir/build
/usr/bin/make  -f library/CMakeFiles/clFFT.dir/build.make library/CMakeFiles/clFFT.dir/build
make[2]: Entering directory '/workspace/srcdir/clFFT/build'
make[2]: Entering directory '/workspace/srcdir/clFFT/build'
[  5%] Building CXX object statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.CPU.cpp.o
[ 10%] Building CXX object statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.GPU.cpp.o
[ 15%] Building CXX object library/CMakeFiles/clFFT.dir/transform.cpp.o
cd /workspace/srcdir/clFFT/build/statTimer && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DSTATTIMER_EXPORTS -DStatTimer_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/statTimer/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.CPU.cpp.o -MF CMakeFiles/StatTimer.dir/statisticalTimer.CPU.cpp.o.d -o CMakeFiles/StatTimer.dir/statisticalTimer.CPU.cpp.o -c /workspace/srcdir/clFFT/src/statTimer/statisticalTimer.CPU.cpp
[ 25%] Building CXX object library/CMakeFiles/clFFT.dir/accessors.cpp.o
[ 25%] Building CXX object statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.extern.cpp.o
cd /workspace/srcdir/clFFT/build/statTimer && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DSTATTIMER_EXPORTS -DStatTimer_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/statTimer/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.GPU.cpp.o -MF CMakeFiles/StatTimer.dir/statisticalTimer.GPU.cpp.o.d -o CMakeFiles/StatTimer.dir/statisticalTimer.GPU.cpp.o -c /workspace/srcdir/clFFT/src/statTimer/statisticalTimer.GPU.cpp
[ 30%] Building CXX object statTimer/CMakeFiles/StatTimer.dir/stdafx.cpp.o
cd /workspace/srcdir/clFFT/build/library && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DCLFFT_EXPORTS -DclFFT_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/library/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT library/CMakeFiles/clFFT.dir/transform.cpp.o -MF CMakeFiles/clFFT.dir/transform.cpp.o.d -o CMakeFiles/clFFT.dir/transform.cpp.o -c /workspace/srcdir/clFFT/src/library/transform.cpp
cd /workspace/srcdir/clFFT/build/library && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DCLFFT_EXPORTS -DclFFT_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/library/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT library/CMakeFiles/clFFT.dir/accessors.cpp.o -MF CMakeFiles/clFFT.dir/accessors.cpp.o.d -o CMakeFiles/clFFT.dir/accessors.cpp.o -c /workspace/srcdir/clFFT/src/library/accessors.cpp
cd /workspace/srcdir/clFFT/build/statTimer && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DSTATTIMER_EXPORTS -DStatTimer_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/statTimer/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.extern.cpp.o -MF CMakeFiles/StatTimer.dir/statisticalTimer.extern.cpp.o.d -o CMakeFiles/StatTimer.dir/statisticalTimer.extern.cpp.o -c /workspace/srcdir/clFFT/src/statTimer/statisticalTimer.extern.cpp
[ 35%] Building CXX object library/CMakeFiles/clFFT.dir/repo.cpp.o
[ 40%] Building CXX object library/CMakeFiles/clFFT.dir/plan.cpp.o
cd /workspace/srcdir/clFFT/build/statTimer && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DSTATTIMER_EXPORTS -DStatTimer_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/statTimer/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT statTimer/CMakeFiles/StatTimer.dir/stdafx.cpp.o -MF CMakeFiles/StatTimer.dir/stdafx.cpp.o.d -o CMakeFiles/StatTimer.dir/stdafx.cpp.o -c /workspace/srcdir/clFFT/src/statTimer/stdafx.cpp
cd /workspace/srcdir/clFFT/build/library && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DCLFFT_EXPORTS -DclFFT_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/library/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT library/CMakeFiles/clFFT.dir/repo.cpp.o -MF CMakeFiles/clFFT.dir/repo.cpp.o.d -o CMakeFiles/clFFT.dir/repo.cpp.o -c /workspace/srcdir/clFFT/src/library/repo.cpp
cd /workspace/srcdir/clFFT/build/library && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DCLFFT_EXPORTS -DclFFT_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/library/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT library/CMakeFiles/clFFT.dir/plan.cpp.o -MF CMakeFiles/clFFT.dir/plan.cpp.o.d -o CMakeFiles/clFFT.dir/plan.cpp.o -c /workspace/srcdir/clFFT/src/library/plan.cpp
/workspace/srcdir/clFFT/src/statTimer/stdafx.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
/workspace/srcdir/clFFT/src/statTimer/statisticalTimer.extern.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
/workspace/srcdir/clFFT/src/library/transform.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
/workspace/srcdir/clFFT/src/statTimer/statisticalTimer.CPU.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
make[2]: *** [statTimer/CMakeFiles/StatTimer.dir/build.make:104: statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.extern.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [statTimer/CMakeFiles/StatTimer.dir/build.make:118: statTimer/CMakeFiles/StatTimer.dir/stdafx.cpp.o] Error 1
make[2]: *** [statTimer/CMakeFiles/StatTimer.dir/build.make:76: statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.CPU.cpp.o] Error 1
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:76: library/CMakeFiles/clFFT.dir/transform.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/workspace/srcdir/clFFT/src/library/repo.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:118: library/CMakeFiles/clFFT.dir/repo.cpp.o] Error 1
/workspace/srcdir/clFFT/src/statTimer/statisticalTimer.GPU.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
/workspace/srcdir/clFFT/src/library/plan.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
/workspace/srcdir/clFFT/src/library/accessors.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
make[2]: *** [statTimer/CMakeFiles/StatTimer.dir/build.make:90: statTimer/CMakeFiles/StatTimer.dir/statisticalTimer.GPU.cpp.o] Error 1
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:104: library/CMakeFiles/clFFT.dir/plan.cpp.o] Error 1
make[2]: Leaving directory '/workspace/srcdir/clFFT/build'
make[1]: *** [CMakeFiles/Makefile2:160: statTimer/CMakeFiles/StatTimer.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:90: library/CMakeFiles/clFFT.dir/accessors.cpp.o] Error 1
[ 50%] Building CXX object library/CMakeFiles/clFFT.dir/generator.transpose.gcn.cpp.o
[ 50%] Building CXX object library/CMakeFiles/clFFT.dir/generator.stockham.cpp.o
cd /workspace/srcdir/clFFT/build/library && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DCLFFT_EXPORTS -DclFFT_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/library/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT library/CMakeFiles/clFFT.dir/generator.stockham.cpp.o -MF CMakeFiles/clFFT.dir/generator.stockham.cpp.o.d -o CMakeFiles/clFFT.dir/generator.stockham.cpp.o -c /workspace/srcdir/clFFT/src/library/generator.stockham.cpp
cd /workspace/srcdir/clFFT/build/library && /opt/bin/i686-linux-gnu-libgfortran3-cxx11/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/ -DCLFFT_EXPORTS -DclFFT_EXPORTS -I/workspace/srcdir/clFFT/build/include -I/workspace/srcdir/clFFT/src/library/../include -m64 -pthread  -O3 -DNDEBUG -fPIC -MD -MT library/CMakeFiles/clFFT.dir/generator.transpose.gcn.cpp.o -MF CMakeFiles/clFFT.dir/generator.transpose.gcn.cpp.o.d -o CMakeFiles/clFFT.dir/generator.transpose.gcn.cpp.o -c /workspace/srcdir/clFFT/src/library/generator.transpose.gcn.cpp
/workspace/srcdir/clFFT/src/library/generator.transpose.gcn.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
/workspace/srcdir/clFFT/src/library/generator.stockham.cpp:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* ************************************************************************
 
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:146: library/CMakeFiles/clFFT.dir/generator.transpose.gcn.cpp.o] Error 1
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:132: library/CMakeFiles/clFFT.dir/generator.stockham.cpp.o] Error 1
make[2]: Leaving directory '/workspace/srcdir/clFFT/build'
make[1]: *** [CMakeFiles/Makefile2:134: library/CMakeFiles/clFFT.dir/all] Error 2
make[1]: Leaving directory '/workspace/srcdir/clFFT/build'
make: *** [Makefile:156: all] Error 2
Previous command exited with 2
ERROR: LoadError: Build for CLFFT on i686-linux-gnu did not complete successfully

Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] autobuild(dir::AbstractString, src_name::AbstractString, src_version::VersionNumber, sources::Vector{<:BinaryBuilderBase.AbstractSource}, script::AbstractString, platforms::Vector, products::Vector{<:Product}, dependencies::Vector{<:BinaryBuilderBase.AbstractDependency}; verbose::Bool, debug::Bool, skip_audit::Bool, ignore_audit_errors::Bool, autofix::Bool, code_dir::Union{Nothing, String}, require_license::Bool, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
   @ BinaryBuilder ~/.julia/packages/BinaryBuilder/wohhx/src/AutoBuild.jl:872
 [3] build_tarballs(ARGS::Any, src_name::Any, src_version::Any, sources::Any, script::Any, platforms::Any, products::Any, dependencies::Any; julia_compat::String, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
   @ BinaryBuilder ~/.julia/packages/BinaryBuilder/wohhx/src/AutoBuild.jl:344
 [4] top-level scope
   @ ~/Desktop/cross-platform-clfft/build_tarballs.jl:47
in expression starting at /home/juliohm/Desktop/cross-platform-clfft/build_tarballs.jl:47
```

**Comment by juliohm** [comment]
Appreciate any help from clFFT maintainers here. We managed to cross-compile OpenCL itself to 14 platforms with the help of the maintainers. We want to do the same for clFFT.

**Comment by juliohm** [comment]
With the patch in #244 the number of supported plataforms went up to 5. Many other platforms are failing with the following error:

```
In file included from /workspace/srcdir/clFFT/src/library/generator.copy.cpp:21:
In file included from /workspace/srcdir/clFFT/src/library/generator.stockham.h:26:
In file included from /workspace/srcdir/clFFT/src/library/repo.h:23:
/workspace/srcdir/clFFT/src/library/plan.h:602:4: warning: delete called on 'FFTAction' that is abstract but has non-virtual destructor [-Wdelete-abstract-non-virtual-dtor]
                        delete action;
                        ^
In file included from /workspace/srcdir/clFFT/src/library/lifetime.cpp:23:
In file included from /workspace/srcdir/clFFT/src/library/repo.h:23:
/workspace/srcdir/clFFT/src/library/plan.h:602:4: warning: delete called on 'FFTAction' that is abstract but has non-virtual destructor [-Wdelete-abstract-non-virtual-dtor]
                        delete action;
                        ^
In file included from /workspace/srcdir/clFFT/src/library/lifetime.cpp:24:
/workspace/srcdir/clFFT/src/library/../include/sharedLibrary.h:64:10: error: "unsupported platform"
        #error "unsupported platform"
         ^
3 warnings and 2 errors generated.
/workspace/srcdir/clFFT/src/library/../include/sharedLibrary.h:67:9: error: use of undeclared identifier 'fileHandle'
        return fileHandle;
               ^
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:160: library/CMakeFiles/clFFT.dir/generator.transpose.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
1 warning and 2 errors generated.
make[2]: *** [library/CMakeFiles/clFFT.dir/build.make:202: library/CMakeFiles/clFFT.dir/lifetime.cpp.o] Error 1
1 warning generated.
In file included from /workspace/srcdir/clFFT/src/library/fft_binary_lookup.cpp:21:
In file included from /workspace/srcdir/clFFT/src/library/fft_binary_lookup.h:30:
/workspace/srcdir/clFFT/src/library/plan.h:602:4: warning: delete called on 'FFTAction' that is abstract but has non-virtual destructor [-Wdelete-abstract-non-virtual-dtor]
                        delete action;
                        ^
10 warnings generated.
1 warning generated.
1 warning generated.
8 warnings generated.
1 warning generated.
make[2]: Leaving directory '/workspace/srcdir/clFFT/build'
make[1]: *** [CMakeFiles/Makefile2:134: library/CMakeFiles/clFFT.dir/all] Error 2
make[1]: Leaving directory '/workspace/srcdir/clFFT/build'
make: *** [Makefile:156: all] Error 2
Previous command exited with 2
```

Appreciate any help.

**Comment by juliohm** [comment]
You can reproduce the error on different platforms.

# FreeBSD

```
$ julia --project build_tarballs.jl --debug x86_64-unknown-freebsd
```

# MacOS

```
$ julia --project build_tarballs.jl --debug x86_64-apple-darwin
```

These commands will drop you in a session that emulates the target platform. The two examples above have the same error. If you can help me fix it, that would be great.

**Comment by juliohm** [comment]
Ok, I fixed the issue on MacOS with #245. Can you please confirm that FreeBSD is not supported? I get some compilation errors saying that the platform is not supported, and would like to double check.
